### PR TITLE
transaction/block format restructure

### PR DESCRIPTION
**Pull request by idandash**

### CODE REVIEW

File: projects/architecture/interfaces/primitives/addresses.proto [hunk 1]
```diff
@@ -0,0 +1,8 @@
+syntax = "proto3";
+
+message UniversalAddress {
+  uint32 scheme = 1;
+  uint32 networkId = 2;
+  bytes id = 3;
+  uint32 checksum = 4;
+}
```
Comment (lbeder) [inline projects/architecture/interfaces/primitives/addresses.proto]: 1. Why do we need so many bytes for `scheme` or `networkId`? 
2. In general, we must to have a way to serialize/deserialize these messages in a fully binary compatible way (e.g., so that we won’t different hashes or signatures). 
3. Why is the checksum a `uint32`? In other words, are we confident that 4 bytes (e.g., 4 bytes of the hash of the address) are enough?

File: projects/architecture/interfaces/primitives/transactions.proto [hunk 2]
```diff
@@ -1,15 +1,32 @@
...
 }
 
-message TransactionAppendix {
-  uint32 version = 1;
-  repeated string prefetchAddresses = 2;
-  string subscriptionKey = 3;
+message TranasactionLifespan {
+  bytes refBlockHash = 1;
+  uint32 blocksToLive = 2;
+}
```
Comment (lbeder) [inline projects/architecture/interfaces/primitives/transactions.proto]: Should be assume single hashing algorithm for all messages (e.g., `blockHash`) or would want to support different hashes?

File: projects/architecture/interfaces/primitives/transactions.proto [hunk 3]
```diff
@@ -1,15 +1,32 @@
...
+
+message TransactionBody {
+  bytes contractAddress = 1;
+  bytes payload = 2;
+}
+
+message Transaction {
+  TransactionHeader header = 1;
+  TransactionBody body = 2;
+  bytes signature = 3;
```
Comment (lbeder) [inline projects/architecture/interfaces/primitives/transactions.proto]: Can we assume single signature scheme (e.g., that the algorithm will be derived from the address)?

File: .vscode/launch.json [hunk 4]
```diff
@@ -0,0 +1,80 @@
+{
```
Comment (lbeder) [inline .vscode/launch.json]: Superb! 🏆

File: e2e/src/orbs-client.ts [hunk 5]
```diff
@@ -57,25 +55,29 @@ export class OrbsClientSession {
   }
 
   async call(contractAddress: string, payload: string) {
-    const { resultJson } = await this.orbsClient.call({
+    const { resultJson } = await this.orbsClient.callContract({
       sender: this.getAddress(),
-      contractAddress: contractAddress,
+      contractAddress: {address: contractAddress},
```
Comment (lbeder) [inline e2e/src/orbs-client.ts]: Why do we need another level of indirection here? I.e., why not `contractAddress,`?

File: e2e/src/orbs-client.ts [hunk 6]
```diff
@@ -57,25 +55,29 @@ export class OrbsClientSession {
...
     return {
-      version: 0,
-      sender: this.senderAddress,
-      contractAddress: contractAddress,
-      payload: payload,
-      signature: "" // TODO: add a signature once implement in the network-side
+      header: {
+        version: 0,
+        sender: this.getAddress(),
+        sequenceNumber: 0
```
Comment (lbeder) [inline e2e/src/orbs-client.ts]: The `sequence` here is confusing, compared how it's used in other blockchains (e.g., Ethereum, Stellar, Ripple, etc.). Maybe something with `order` instead?

File: projects/architecture/interfaces/primitives/addresses.proto [hunk 7]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+
+message UniversalAddress {
```
Comment (lbeder) [inline projects/architecture/interfaces/primitives/addresses.proto]: This will change, according to the last discussions with architecture. I'll address its affects on the SDK and UniversalAddress in a future PR.

File: projects/architecture/interfaces/primitives/addresses.proto [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+  uint32 scheme = 1;
+  uint32 networkId = 2;
+  bytes id = 3; // TODO: is id the right name? maybe key?
```
Comment (lbeder) [inline projects/architecture/interfaces/primitives/addresses.proto]: Today, we refer to it as `accountId`.

File: projects/architecture/interfaces/primitives/addresses.proto [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+  uint32 checksum = 4;
+}
+
+message ContractAddress {
+  string address = 1; // TODO: what's the format?
```
Comment (lbeder) [inline projects/architecture/interfaces/primitives/addresses.proto]: IMHO, it'll be a `UniversalAddress` as well, but now it's not the time anyway.

File: projects/architecture/interfaces/primitives/blocks.proto [hunk 10]
```diff
@@ -1,21 +1,27 @@
...
 message BlockHeader {
   uint32 version = 1;
-  uint32 id = 2;
-  uint32 prevBlockId = 3;
+  bytes prevBlockHash = 2;
+  uint32 height = 3;
+}
+
+message BlockBody {
+  repeated Transaction transactions = 2;
```
Comment (lbeder) [inline projects/architecture/interfaces/primitives/blocks.proto]: Let's start from 1, while we still can?

File: projects/architecture/interfaces/primitives/blocks.proto [hunk 11]
```diff
@@ -1,21 +1,27 @@
...
   string key = 2;
   string value = 3;
 }
 
 message Block {
   BlockHeader header = 1;
-  repeated Transaction transactions = 2;
-  repeated ModifiedStateKey stateDiff = 3;
+  BlockBody body = 2;
+  bytes hash = 3; // TODO: how is the hash calculated?
```
Comment (lbeder) [inline projects/architecture/interfaces/primitives/blocks.proto]: Do we need a hash here at all? It can be quickly derived from everything (e.g., there is no block hash in either Bitcoin or Ethereum).

File: projects/architecture/interfaces/primitives/transactions.proto [hunk 12]
```diff
@@ -1,15 +1,36 @@
...
+  // bytes signature = 3;
 }
 
-message TransactionAppendix {
+message TransactionHeader {
   uint32 version = 1;
-  repeated string prefetchAddresses = 2;
-  string subscriptionKey = 3;
+  UniversalAddress sender = 2;
+  // TransactionLifespan lifespan = 3;
```
Comment (lbeder) [inline projects/architecture/interfaces/primitives/transactions.proto]: Why have you decided to not address lifespan/ttl here? If it's going to be part of a different PR, could you remove this commented out code from here for now?

File: projects/libs/core-library-typescript/src/block-storage/block-storage.ts [hunk 13]
```diff
@@ -59,56 +65,52 @@ export class BlockStorage {
...
-      throw new Error(`Invalid block ID of block: ${JSON.stringify(block)}!`);
+    if (block.header.height !== this.lastBlock.header.height + 1) {
+      throw new Error(`Invalid block height of block: ${JSON.stringify(block)}!`);
     }
 
-    if (block.header.prevBlockId !== this.lastBlock.header.id) {
-      throw new Error(`Invalid prev block ID of block: ${JSON.stringify(block)}! Should have been ${this.lastBlock.header.id}`);
+    if (!block.header.prevBlockHash.equals(this.lastBlock.hash)) {
+      throw new Error(`Invalid prev block hash of block: ${JSON.stringify(block)}! Should have been ${JSON.stringify(this.lastBlock.hash)}`);
     }
```
Comment (lbeder) [inline projects/libs/core-library-typescript/src/block-storage/block-storage.ts]: I don't think we should have a hash in the block at all, but if we will - its verifisation should be here as well.

File: projects/libs/core-library-typescript/src/block-storage/block-storage.ts [hunk 14]
```diff
@@ -59,56 +65,52 @@ export class BlockStorage {
...
   }
 
-  public async getBlock(id: number): Promise<types.Block> {
-    return JSON.parse(await this.db.get<string>(id.toString()));
+  public async getBlock(height: number): Promise<types.Block> {
+    return JsonBuffer.parseJsonWithBuffers(await this.db.get<string>(height.toString()));
```
Comment (lbeder) [inline projects/libs/core-library-typescript/src/block-storage/block-storage.ts]: 👍

File: projects/libs/core-library-typescript/src/common-library/block-utils.ts [hunk 15]
```diff
@@ -0,0 +1,38 @@
+
```
Comment (lbeder) [inline projects/libs/core-library-typescript/src/common-library/block-utils.ts]: Extra line

File: projects/libs/core-library-typescript/src/common-library/block-utils.ts [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+import { types, logger } from ".";
+import { createHash } from "crypto";
+
+export namespace BlockUtils {
+  export function calculateBlockHash(blockComponents: {header: types.BlockHeader, body: types.BlockBody}): Buffer {
+    const hash = createHash("sha256");
```
Comment (lbeder) [inline projects/libs/core-library-typescript/src/common-library/block-utils.ts]: It's common to do a double-sha256 here (i.e., `hash = sha256(sha256(block))`) as a "better safe than sorry" mitigation against a similar attack on sha1. Alternatively, we can do sha3 (keccak) once instead.

File: projects/libs/core-library-typescript/src/common-library/block-utils.ts [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+    return block;
+  }
+
+  export function buildNextBlock(body: types.BlockBody, prevBlock?: types.Block): types.Block {
+    return buildBlock({
+      header: {
+        version: 0,
+        prevBlockHash: prevBlock ? prevBlock.hash : new Buffer(""),
+        height: prevBlock ? prevBlock.header.height + 1 : 0
```
Comment (lbeder) [inline projects/libs/core-library-typescript/src/common-library/block-utils.ts]: This should error. It shouldn't be possible to build a block with `height` 0 - this will always be the Genesis block, so if `prevBlock` is undefined - this method should error.

File: projects/libs/core-library-typescript/src/common-library/block-utils.ts [hunk 18]
```diff
@@ -0,0 +1,38 @@
+
+import { types, logger } from ".";
+import { createHash } from "crypto";
+
+export namespace BlockUtils {
+  export function calculateBlockHash(blockComponents: {header: types.BlockHeader, body: types.BlockBody}): Buffer {
```
Comment (lbeder) [inline projects/libs/core-library-typescript/src/common-library/block-utils.ts]: Instead of a namespace + flat function, could you implement it as a class with static methods instead? It'd be more similar to what we did in other places (and what we will do in other languages).

File: projects/libs/core-library-typescript/src/consensus/block-builder.ts [hunk 19]
```diff
@@ -1,36 +1,103 @@
...
 
-    public async buildNextBlock(lastBlockId: number) {
-        const { transactions } = await this.transactionPool.getAllPendingTransactions({});
+    const { processedTransactions, stateDiff } = await this.virtualMachine.processTransactionSet({ orderedTransactions: transactions });
 
-        if (transactions.length == 0) {
-            throw new Error("Transaction pool is empty");
-        }
+    if (processedTransactions.length == 0) {
+        throw new Error("Transaction pool is empty");
```
Comment (lbeder) [inline projects/libs/core-library-typescript/src/consensus/block-builder.ts]: @netoneko discovered an issue with this approach: unprocessed transactions (regardless if all or only some) aren't removed from the transaction pool(s). Could you address it as well?

File: projects/libs/core-library-typescript/src/consensus/block-builder.ts [hunk 20]
```diff
@@ -1,36 +1,103 @@
...
+    const block = await this.buildBlockFromPendingTransactions(lastBlock);
+
+    this.onNewBlockBuild(block);
+
+    logger.debug(`appended new block ${JSON.stringify(block)}`);
+
+    this.readyForBlockAppend = false;
+  }
+
+
```
Comment (lbeder) [inline projects/libs/core-library-typescript/src/consensus/block-builder.ts]: Extra line

File: projects/libs/core-library-typescript/src/consensus/raft-consensus.ts [hunk 21]
```diff
@@ -7,6 +7,9 @@ import BlockBuilder from "./block-builder";
 
 import { Gossip } from "../gossip";
 import { Block } from "web3/types";
+import { build } from "protobufjs";
+import { SSL_OP_CRYPTOPRO_TLSEXT_BUG } from "constants";
```
Comment (lbeder) [inline projects/libs/core-library-typescript/src/consensus/raft-consensus.ts]: What's that?

File: projects/libs/core-library-typescript/src/state-storage/state-storage.ts [hunk 22]
```diff
@@ -5,64 +5,67 @@ import { types } from "../common-library/types";
...
-    return new Promise((resolve, reject) => {
-      if (blockId > this.lastBlockId) {
-        if (timeout < 200) {
-          reject(new Error(`Timeout in attempt to read block state (${blockId} != ${this.lastBlockId})`));
-        } else {
-          setTimeout(() => this.waitForBlockState(timeout - 200), 200);
-        }
+    const { block } = await this.blockStorage.getLastBlock({});
+    while (timeout > 0) {
+        if (block.header.height <= this.lastBlockHeight) {
```
Comment (lbeder) [inline projects/libs/core-library-typescript/src/state-storage/state-storage.ts]: Indentation

File: projects/libs/core-library-typescript/test/block-storage/block-storage-sync.spec.ts [hunk 23]
```diff
@@ -4,111 +4,99 @@ import * as chaiAsPromised from "chai-as-promised";
...
-  return {
-    header: {
-      version: 0,
-      id: prevBlockId + 1,
-      prevBlockId: prevBlockId
-    },
-    tx: { version: 0, contractAddress: "0", sender: "", signature: "", payload: "{}" },
-    modifiedAddressesJson: "{}"
-  };
+function generateNextNullBlock(lastBlock: types.Block): types.Block {
```
Comment (lbeder) [inline projects/libs/core-library-typescript/test/block-storage/block-storage-sync.spec.ts]: Maybe `EmptyBlock`?

File: projects/libs/core-library-typescript/test/block-storage/block-storage-sync.spec.ts [hunk 24]
```diff
@@ -4,111 +4,99 @@ import * as chaiAsPromised from "chai-as-promised";
...
       blockStorageSync.onReceiveBlock(blocks[3]);
 
-      blockStorageSync.getQueueSize().should.be.eql(5);
-
-      const appendBlocksPromise = blockStorageSync.appendBlocks();
-      try {
-        await appendBlocksPromise;
-      } catch (e) {
-      }
+      blockStorageSync.getQueueSize().should.eql(5);
```
Comment (lbeder) [inline projects/libs/core-library-typescript/test/block-storage/block-storage-sync.spec.ts]: Let's try to stick to `expect`? There is no benefit of using both at the same time all around the project :(

File: projects/libs/core-library-typescript/test/consensus/block-builder.spec.ts [hunk 25]
```diff
@@ -1,76 +1,109 @@
-import { types } from "../../src/common-library/types";
+import { types,  } from "../../src/common-library/types";
```
Comment (lbeder) [inline projects/libs/core-library-typescript/test/consensus/block-builder.spec.ts]: Extra comma

File: e2e/src/orbs-client.ts (hunk 5 above)
Comment (idandash) [reply e2e/src/orbs-client.ts]: It's not. It was just an abstraction layer I added as an interim step towards a new structure, while still keeping the existing addressing scheme

File: e2e/src/orbs-client.ts (hunk 6 above)
Comment (idandash) [reply e2e/src/orbs-client.ts]: Maybe. Let's wait for the final spec and change accordingly

File: projects/libs/core-library-typescript/src/common-library/block-utils.ts (hunk 17 above)
Comment (idandash) [reply projects/libs/core-library-typescript/src/common-library/block-utils.ts]: It was added for testing purposes

File: projects/libs/core-library-typescript/src/consensus/raft-consensus.ts (hunk 21 above)
Comment (idandash) [reply projects/libs/core-library-typescript/src/consensus/raft-consensus.ts]: Oops..

File: projects/libs/core-library-typescript/test/block-storage/block-storage-sync.spec.ts (hunk 23 above)
Comment (idandash) [reply projects/libs/core-library-typescript/test/block-storage/block-storage-sync.spec.ts]: 👍

File: projects/libs/core-library-typescript/test/block-storage/block-storage-sync.spec.ts (hunk 24 above)
Comment (idandash) [reply projects/libs/core-library-typescript/test/block-storage/block-storage-sync.spec.ts]: I do, it's just that this module was already using should() and I didn't want to touch it too much

File: projects/libs/core-library-typescript/src/common-library/block-utils.ts (hunk 16 above)
Comment (idandash) [reply projects/libs/core-library-typescript/src/common-library/block-utils.ts]: It's more a placeholder until the hashing scheme is finalized

File: projects/architecture/interfaces/primitives/transactions.proto (hunk 12 above)
Comment (idandash) [reply projects/architecture/interfaces/primitives/transactions.proto]: It's a placeholder. It was a reasonable compromise since I do aim to implement it very soon
